### PR TITLE
GitHub CI: Test with AddressSanitizer

### PR DESCRIPTION
**Pull request by masterleinad**
Related to #6151 and #6674. AddressSanitizer would have detected problems in #6151 without #6674 (if only the `Serial` backend was enabled). Since we are treating atomics special if only the `Serial` backend was enabled (also see #6672), I added a build with AddressSanitizer that only enables `Serial`. I also enabled it for another build (with the `Threads` backend) for which it was convenient.

### CODE REVIEW

**Comment by masterleinad** [comment]
All GitHub CI passes.

**Review by ldh4 — APPROVED** [review]
I'm not sure about needing this for `threads`, but not particularly against it either.

File: .github/workflows/continuous-integration-workflow.yml [hunk 1]
```diff
@@ -115,9 +124,12 @@ jobs:
           cmake --build builddir --parallel 2
           ccache -s
       - name: Tests
+        env:
+          ASAN_OPTIONS: allocator_may_return_null=1
```
Comment (dalg24) [inline .github/workflows/continuous-integration-workflow.yml:128]: What is this option needed for?

Comment (masterleinad) [reply .github/workflows/continuous-integration-workflow.yml:128]: We intentionally allocate more memory than reasonably available to check the error message in some tests. It seemed reasonable to not care too much about this particular check.

Comment (masterleinad) [reply .github/workflows/continuous-integration-workflow.yml]: See [`09f12c6` (#6676)](https://github.com/kokkos/kokkos/pull/6676/commits/09f12c6983657b606489b5a824918a8c02e71c3d).

Comment (dalg24) [reply .github/workflows/continuous-integration-workflow.yml]: Why did you not use an attribute?
https://clang.llvm.org/docs/AddressSanitizer.html#disabling-instrumentation-with-attribute-no-sanitize-address

Comment (masterleinad) [reply .github/workflows/continuous-integration-workflow.yml]: I'm not quite sure how exactly I would use it for just this unit test and disabling the unit test seemed good enough.

Comment (masterleinad) [reply .github/workflows/continuous-integration-workflow.yml]: Looks like the attempt in 335f7f2  for using attributes doesn't really work.

Comment (dalg24) [reply .github/workflows/continuous-integration-workflow.yml]: I never managed to get it to work when I tried on my Mac a while ago.  Thanks for trying anyway.

**Review by dalg24 — APPROVED** [review]

**Pull request merged**